lib/actions-oclif/config/generate.ts

```diff
@@ -0,0 +1,237 @@
+			description: 'device uuid',
+			char: 'd',
+			exclusive: ['application', 'app'],
+		}),
+		deviceApiKey: flags.string({
+			description:
+				'custom device key - note that this is only supported on balenaOS 2.0.3+',
+			char: 'k',
+		}),
+		deviceType: flags.string({
+			description: 'device type slug',
+		}),
+		'generate-device-api-key': flags.boolean({
+			description: 'generate a fresh device key for the device',
+		}),
+		output: flags.string({
+			description: 'path of output file',
+			char: 'o',
+		}),
+		// Options for non-interactive configuration
+		network: flags.string({
+			description: 'the network type to use: ethernet or wifi',
+			options: ['ethernet', 'wifi'],
+		}),
+		wifiSsid: flags.string({
+			description:
+				'the wifi ssid to use (used only if --network is set to wifi)',
+		}),
+		wifiKey: flags.string({
+			description:
+				'the wifi key to use (used only if --network is set to wifi)',
+		}),
+		appUpdatePollInterval: flags.string({
+			description:
+				'how frequently (in minutes) to poll for application updates',
+		}),
+		help: cf.help,
+	};
+
+	public static authenticated = true;
+
+	public async run() {
+		const { flags: options } = this.parse<FlagsDef, {}>(ConfigGenerateCmd);
+
+		const balena = getBalenaSdk();
+
+		await this.validateOptions(options);
+
+		// Get device | application
+		let resource;
+		if (options.device != null) {
+			const { tryAsInteger } = await import('../../utils/validation');
+			resource = await balena.models.device.get(tryAsInteger(options.device));
+		} else {
+			resource = await balena.models.application.get(options.application!);
+		}
+
+		const deviceType = options.deviceType || resource.device_type;
+
+		const deviceManifest = await balena.models.device.getManifestBySlug(
+			deviceType,
+		);
+
+		// Check compatibility if application and deviceType provided
+		if (options.application && options.deviceType) {
+			const appDeviceManifest = await balena.models.device.getManifestBySlug(
+				resource.device_type,
+			);
+
+			const helpers = await import('../../utils/helpers');
+			if (
+				!helpers.areDeviceTypesCompatible(appDeviceManifest, deviceManifest)
+			) {
+				throw new balena.errors.BalenaInvalidDeviceType(
+					`Device type ${options.deviceType} is incompatible with application ${options.application}`,
+				);
+			}
+		}
+
+		// Prompt for values
+		// Pass params as an override: if there is any param with exactly the same name as a
+		// required option, that value is used (and the corresponding question is not asked)
+		const answers = await getCliForm().run(deviceManifest.options, {
+			override: options,
+		});
+		answers.version = options.version;
+
+		// Generate config
+		const { generateDeviceConfig, generateApplicationConfig } = await import(
+			'../../utils/config'
+		);
+
+		let config;
+		if ('uuid' in resource && resource.uuid != null) {
+			config = await generateDeviceConfig(
+				resource as Device & { belongs_to__application: PineDeferred },
```

Comment:
Could you move this above as
```ts
resource = await balena.models.device.get(tryAsInteger(options.device)) as Device & { belongs_to__application: PineDeferred };
```
that way this can be just
```suggestion
				resource,
```
and the actual cast will happen at the point we fetch the device and make it easier to verify that the cast is valid, and also mean that all usage of the `Device` variant of `resource` will be typed more accurately

Reply:
Yeah that's a bit nicer.